crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -25,61 +25,119 @@ pub struct Fulfillment {
     pub executed: competition::order::TargetAmount,
 }
 
-impl Fulfillment {
+/// A trade which adds a JIT order. See [`order::Jit`].
+#[derive(Debug)]
+pub struct Jit {
+    pub order: order::Jit,
+    pub executed: competition::order::TargetAmount,
+}
+
+impl Trade {
     /// Calculate the settlement contract input and output amounts executed by
     /// this trade.
     pub fn execution(&self, clearing_prices: &ClearingPrices) -> Result<Execution, Error> {
-        let input = self.executed.to_asset(&self.order);
-        let (input_price, output_price) = {
-            let sell_price = clearing_prices
-                .0
-                .get(&self.order.sell.token)
-                .ok_or(Error::ClearingPriceMissing)?
-                .to_owned();
-            let buy_price = clearing_prices
-                .0
-                .get(&self.order.buy.token)
-                .ok_or(Error::ClearingPriceMissing)?
-                .to_owned();
-            match self.order.side {
-                order::Side::Buy => (buy_price, sell_price),
-                order::Side::Sell => (sell_price, buy_price),
-            }
-        };
-        let output = eth::Asset {
-            amount: match self.order.kind {
-                order::Kind::Market => input
-                    .amount
-                    .checked_mul(input_price)
-                    .ok_or(Error::Overflow)?
-                    .checked_ceil_div(&output_price)
-                    .ok_or(Error::Overflow)?,
-                order::Kind::Limit { .. } => todo!(),
-                order::Kind::Liquidity => todo!(),
+        // Values needed to calculate the executed amounts.
+        let ExecutionParams {
+            side,
+            kind,
+            sell,
+            buy,
+            executed,
+        } = match self {
+            Trade::Fulfillment(trade) => ExecutionParams {
+                side: trade.order.side,
+                kind: trade.order.kind,
+                sell: trade.order.sell,
+                buy: trade.order.buy,
+                executed: trade.executed,
             },
-            token: match self.order.side {
-                order::Side::Buy => self.order.sell.token,
-                order::Side::Sell => self.order.buy.token,
+            Trade::Jit(trade) => ExecutionParams {
+                side: trade.order.side,
+                // For the purposes of calculating the executed amounts, a JIT order behaves the
+                // same as a regular market order.
+                kind: order::Kind::Market,
```

Comment:
Jit orders are supposed to be liquidity orders. These orders don't adhere to the uniform clearing prices and always settle at their limit price.

For fill-or-kill JIT orders you can simply take the `sell` and `buy` amount as is. And for partially fillables you have to compute the filled percentage with the `executed_amount`, `side`, `sell`, `buy` and scale the `sell` and `buy` values accordingly.